java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedProcessFunction.java

```diff
@@ -96,93 +142,211 @@ public void onTimer(
+                return;
+            }
+            expiredRowHandler.handleExpiredRow(out, lastRow, timestamp);
+            state.lastOutputRow.clear();
             return;
         }
 
-        final Row aggResultRow =
-                Row.of(
-                        aggregationFieldsDescriptor.getAggFieldDescriptors().stream()
-                                .map(d -> d.aggFunc.getResult())
-                                .toArray());
-
-        out.collect(
-                Row.join(
-                        ctx.getCurrentKey(),
-                        aggResultRow,
-                        Row.of(Instant.ofEpochMilli(timestamp))));
+        state.lastOutputRow.update(outputRow);
+        out.collect(outputRow);
     }
 
     /** The state of {@link SlidingWindowKeyedProcessFunction}. */
-    private static class MultiWindowSizeState {
+    private static class SlidingWindowState {
+
+        /** This MapState maps from row timestamp to the row. */
         private final MapState<Long, Row> timestampToRow;
 
-        // One KeyedProcessFunction can process rows from different keys. Unlike KeyedState, which
-        // is already scoped by key by Flink, we need to have a map that store the timestamp list
-        // for each key.
-        private final Map<Row, LinkedList<Long>> keyToTimestamps;
+        /**
+         * This ListState keeps all the row timestamp that has been added to the timestampToRow
+         * state. Since the operator assume rows are ordered by time, the timestamp list should be
+         * ordered as well.
+         */
+        private final ListState<Long> timestampList;
+
+        /**
+         * This ValueState keeps a map from output field name to the left index of the
+         * timestampList. We use ValueState of Map instead of MapState because we always get indexes
+         * of all aggregation functions. This way we can reduce the read from state backend. If the
+         * index is not null, it always points to the timestamp of the first row in the current
+         * window. If it is null, it means there is no row in the current window.
+         */
+        private final ValueState<Map<String, Integer>> outFieldNameToLeftTimestampIdx;
 
-        private MultiWindowSizeState(MapState<Long, Row> mapState) {
-            this.timestampToRow = mapState;
-            this.keyToTimestamps = new HashMap<>();
+        /**
+         * This ValueState keeps the maximum registered timer so that we don't try to register the
+         * same timer twice.
+         */
+        private final ValueState<Long> maxRegisteredTimer;
+
+        /**
+         * This ValueState keeps the last output row so that we can handle last output row when it
+         * is expired.
+         */
+        private final ValueState<Row> lastOutputRow;
+
+        /**
+         * This is a ValueState of Row that stores the accumulators for each aggregation function.
+         */
+        private final ValueState<Row> accumulatorsRow;
```

Comment:
Typically accumulator has the computation logic for add/retract. Since the returned value is a simple POJO class containing the accumulate value, how about renaming it as `accumulatorStates`?